src/transport/manager/mod.rs

```diff
@@ -625,35 +625,59 @@ impl TransportManager {
         {
             let mut peers = self.peers.write();
 
-            match peers.get_mut(&remote_peer_id) {
-                None => {
-                    drop(peers);
-                    self.peers.write().insert(
-                        remote_peer_id,
-                        PeerContext {
-                            state: PeerState::Dialing {
-                                record: record.clone(),
-                            },
-                            addresses: AddressStore::new(),
-                            secondary_connection: None,
-                        },
+            match peers.entry(remote_peer_id) {
+                Entry::Occupied(occupied) => {
+                    let context = occupied.into_mut();
+
+                    // `context.addresses.insert` does not overwrite the existing record.
+                    // This inserts the record with a score of 0 if it doesn't exist, for
+                    // keeping track of the potential address. This becomes useful for
+                    // consecutive `fn dial` calls.
+                    // However, the record score cannot be updated in the future.
+                    // TODO: context.addresses.insert(record.clone());
```

Comment:
May be this is what you mean already by this TODO, but for me it looks like we have to store multiple `records` at least in `PeerState::Dialing` state to correctly handle concurrent dials (do we dial 8 addresses in parallel?) and track all `connection_id`s. This way the `debug_assert!`s in the `on_dial_failure()` shouldn't be triggered.

May be it's slightly more complicated than this, as we have to track what dials succeeded and drop the extra connections when there is more than two of them, so more bookkeeping might be needed.

> `// However, the record score cannot be updated in the future.`

Can you also elaborate on this?

Reply:
I think the `Opening` state should handle this, but let me double check

Yeaa, the comment is wrong it should be "The record score can be updated in the future" :D
Basically, we can keep track of this potential address in the future, instead of discarding it. So if we ever try to fn dial() 8 addresses, this address can be one of the dialed ones